CMakeLists.txt

```diff
@@ -227,7 +230,6 @@ endif ()
 # Make this extra-checks for correct library dependencies.
 if (NOT SANITIZE)
     set (CMAKE_EXE_LINKER_FLAGS    "${CMAKE_EXE_LINKER_FLAGS} -Wl,--no-undefined")
-    set (CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--no-undefined")
```

Comment:
Why this flag was disabled?

Reply:
I have tried to enable it back and found a lot of broken dependencies. What was the purpose to commit these changes?

Reply:
> @amosbird Why this flag was disabled?

It's to split circular dependencies. 

> I have tried to enable it back and found a lot of broken dependencies. What was the purpose to commit these changes?

The purpose is to split libdbms.so in order to speed up linking. IIUC, static build can make sure all symbols have the correct resolution, and shared build is for developers to quickly iterate. Why do we need `-Wl,--no-undefined` for shared libs?

Reply:
There are two purposes.

1. Check the correctness of implied dependencies. CMake doesn't have visibility resolution or dependency enforcement (i.e. like GN), so we can include some header from another library without linking it properly via CMakeLists.txt - it leads to architecture problems and fails code decomposition.
2. Run shared build locally. I haven't tried but I suggest that `no-undefined` doesn't work transitively and undefined symbols from shared libraries won't be checked from resulting executable. It also leads to a lot of painful fixes for a days, since they are not caught during CI checks anymore.

Reply:
> The purpose is to split libdbms.so in order to speed up linking.
I suggest that the proper way to achieve this goal is to properly refactor and decompose dbms library source code into independent libraries - with explicit source lists and dependencies in between.

Reply:
> the proper way to achieve this goal 

I agree, and I did follow this track at first https://github.com/yandex/ClickHouse/pull/6564 . However, it's very complicated to split further, which also means a huge refactor, with the only gain for developers (linking faster). 

It has been said that shared build is made only for developers, I don't see why undefined symbols is a problem. If it occurs, just link necessary libs back. Static production build won't be affected at all.

Reply:
For the history: decided to implement simple build-and-run scenario for shared build in CI, and remove requirement for `no-undefined` linker option.